docs/adr/0004-only-threadsafe-interfaces.md

```diff
@@ -0,0 +1,118 @@
+# Remove support for non-threadsafe interfaces
+
+* Status: proposed
+* Deciders: mhammond, proposed: rfkelly, jhugman, dmose, travis, janerik
+* Date: 2021-03-31
+
+Technical Story: [Issue 419](https://github.com/mozilla/uniffi-rs/issues/419)
+
+## Context and Problem Statement
+
+[ADR-0003](0003-threadsafe-interfaces.md) introduced support for "thread-safe
+interfaces" - possibly leading to the impression that there is such a thing as
+non-threadsafe interfaces.
+
+However, the entire concept of non-threadsafe interfaces is a mismomer -
+everything wrapped by uniffi is thread-safe - the only question is who manages
+this thread-safety. Interfaces which are not marked as thread-safe cause uniffi
+to wrap the interface in a mutex which is hidden in the generated code and
+therefore not obvious to the casual reader.
+
+## Decision Drivers
+
+* Non-threadsafe structs are considered a "foot-gun", leading to accidentally
```

Comment:
nit: grammer here..."leading us to" or "leading to accidental introduction of" or similar?

Reply:
I think it may be worth saying explicitly that the issue it causes is "method calls that can unexpectedly block for long periods" or similar.